presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -211,17 +218,85 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
 
             Map<Boolean, List<Expression>> conjuncts = extractConjuncts(context.get()).stream().collect(Collectors.partitioningBy(deterministic));
 
-            // Push down conjuncts from the inherited predicate that don't depend on non-deterministic assignments
-            PlanNode rewrittenNode = context.defaultRewrite(node, inlineSymbols(node.getAssignments().getMap(), combineConjuncts(conjuncts.get(true))));
+            // Push down conjuncts from the inherited predicate that only depend on deterministic assignments with
+            // certain limitations. The non-inlining targets are the ones that are left from the symbols of the node
+            List<Expression> deterministicConjuncts = conjuncts.get(true);
+            Sets.SetView<Symbol> nonInliningTargets = extractNonInliningTargets(deterministicConjuncts, node);
 
-            // All non-deterministic conjuncts, if any, will be in the filter node.
+            // We partition the expressions in the deterministicConjuncts into two lists, and only inline the
+            // expressions that are in the inlining targets list.
+            Map<Boolean, List<Expression>> partitionedConjuncts = partitionConjuncts(deterministicConjuncts, nonInliningTargets);
+            List<Expression> inlinedDeterministicConjuncts = partitionedConjuncts.get(true).stream()
+                    .map(entry -> inlineSymbols(node.getAssignments().getMap(), entry))
+                    .collect(Collectors.toList());
+
+            PlanNode rewrittenNode = context.defaultRewrite(node, combineConjuncts(inlinedDeterministicConjuncts));
+
+            // All deterministic conjuncts that contains non-inlining targets, and non-deterministic conjuncts,
+            // if any, will be in the filter node.
+            List<Expression> nonInliningConjuncts = partitionedConjuncts.get(false);
             if (!conjuncts.get(false).isEmpty()) {
-                rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(conjuncts.get(false)));
+                nonInliningConjuncts.addAll(conjuncts.get(false));
             }
+            rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(nonInliningConjuncts));
 
             return rewrittenNode;
         }
 
+        private Sets.SetView<Symbol> extractNonInliningTargets(List<Expression> conjuncts, ProjectNode node)
```

Comment:
I'm replacing the function to return boolean indicating whether an expression should be inlined. The previous implementation was based on InlineProjections and was not right for pushing down filters because it counts the dependencies across all expressions.